pyjoycon/joycon.py

```diff
@@ -1,52 +1,68 @@
 import hid
 import time
 import threading
-import pyjoycon.device as d
 
 
 class JoyCon:
     VENDOR_ID = 0x057E
     L_PRODUCT_ID = 0x2006
     R_PRODUCT_ID = 0x2007
-    _L_ACCEL_OFFSET_X = 350
-    _L_ACCEL_OFFSET_Y = 0
-    _L_ACCEL_OFFSET_Z = 4081
-    _R_ACCEL_OFFSET_X = 350
-    _R_ACCEL_OFFSET_Y = 0
-    _R_ACCEL_OFFSET_Z = -4081
+    _PRESET_L_ACCEL_OFFSET = (350, 0,  4081)
+    _PRESET_R_ACCEL_OFFSET = (350, 0, -4081)
+    _PRESET_L_GYRO_OFFSET = (0, 0, 0)
+    _PRESET_R_GYRO_OFFSET = (0, 0, 0)
 
     _INPUT_REPORT_SIZE = 49
-    _INPUT_REPORT_FREQ = 1.0 / 60.0
+    _INPUT_REPORT_PERIOD = 1.0 / 60.0
     _RUMBLE_DATA = b'\x00\x01\x40\x40\x00\x01\x40\x40'
 
-    def __init__(self, vendor_id, product_id):
+    def __init__(self, vendor_id: int, product_id: int, serial: str = None):
         if vendor_id != self.VENDOR_ID:
             raise ValueError('vendor_id is invalid')
 
-        if product_id not in [self.L_PRODUCT_ID, self.R_PRODUCT_ID]:
+        if product_id not in (self.L_PRODUCT_ID, self.R_PRODUCT_ID):
             raise ValueError('product_id is invalid')
 
-        self._joycon_device = self._open(vendor_id, product_id)
-        self._PRODUCT_ID = product_id
+        # setup internal state
+        self._input_hooks = []
+        self._joycon_device = None
+        self._input_report = bytes(self._INPUT_REPORT_SIZE)
         self._packet_number = 0
+        self._PRODUCT_ID = product_id
+        self._SERIAL_NUMBER = serial
+        if self.is_left():
+            self.set_accel_callibration(self._PRESET_L_ACCEL_OFFSET)
+            self.set_gyro_callibration(self._PRESET_L_GYRO_OFFSET)
+        else:
+            self.set_accel_callibration(self._PRESET_R_ACCEL_OFFSET)
+            self.set_gyro_callibration(self._PRESET_R_GYRO_OFFSET)
+
+        # connect to joycon
+        self._joycon_device = self._open(vendor_id, product_id, serial=None)
         self._setup_sensors()
 
-        self._input_report = bytes(self._INPUT_REPORT_SIZE)
+        # start talking with the joycon in a daemon thread
         self._update_input_report_thread = threading.Thread(
             target=self._update_input_report)
         self._update_input_report_thread.setDaemon(True)
         self._update_input_report_thread.start()
 
-    def _open(self, vendor_id, product_id):
-        _joycon_device = hid.device()
+    def _open(self, vendor_id, product_id, serial):
         try:
-            _joycon_device.open(vendor_id, product_id)
-        except IOError:
-            raise IOError('joycon connect failed')
+            if hasattr(hid, "device"):  # hidapi
+                _joycon_device = hid.device(vendor_id, product_id, serial)
```

Comment:
This doesn't work because the `hid.device` class of the `hidapi` module has no such constructor.
Please check definition of it in [this file](https://github.com/trezor/cython-hidapi/blob/master/hid.pyx).
So please change the initialization process like below.
```suggestion
                _joycon_device = hid.device()
                _joycon_device.open(vendor_id, product_id, serial)
```